Implement periodic quorum shuffling

AndrejMitrovic opened the pull request:
Notes:

- This does not implement saving of the preimages as mentioned in #1069. That's a separate issue. In the tests I simply poll and wait for the preimages to be revealed.
- When the shuffle cycle is set to a lower value (20), the EnrollmentManager tests frequently fail (because they generate around 21 blocks..). I'm pretty certain it's because the tests are written in a wrong way, one example shown was https://github.com/bpfkorea/agora/issues/1071 but there are more tests that need to be fixed. I've set the default shuffle cycle to 30 for now until this is resolved.
- This PR should be split up. It includes 4 bugfix/styling commits, an enhancement by making the quorum generator parameters configurable, and then the actual quorum shuffling implementation. All of the commits are required though.

Fixes: #1069

AndrejMitrovic committed:
Add a quorum shuffle interval consensus parameter

The shuffle interval specifies the number of blocks between
each quorum shuffle takes place. Note that a shuffle may occur
before the cycle ends if the active validator set changes
(new enrollments, expired enrollments..)

AndrejMitrovic committed:
Introduce overridable onAcceptedBlock() method

This method is called when a block was added
to the ledger. The Validator will later override
this and add quorum shuffling logic.

AndrejMitrovic committed:
Implement periodic quorum shuffling

The new QuorumShuffleCycle consensus parameter
sets the maximum number of blocks before a
quorum shuffle takes place.

Note that a shuffle may occur before the cycle
ends if the active validator set changes
(new enrollments, expired enrollments..)

Fixes #1069

AndrejMitrovic commented:
Rebased.

MichaelKim20 commented:
I have some suggestions and write them.
I think introducing a shuffle cycle is a good choice.
I think there should be a clearer standard for the value of the shuffle cycle.
I calculated it and I think up to 40 is possible.
I think it would be good to connect the shuffle cycle with the block height.
Cycle 0: Block Height 0-39
Cycle 1: Block Height 40-79
Cycle 2: Block Height 80-119

AndrejMitrovic commented:
> I think there should be a clearer standard for the value of the shuffle cycle.

It's a configurable option. We can change the default later.

> I think it would be good to connect the shuffle cycle with the block height.

That could cause unnecessary extra shuffles if a new validator has just enrolled.

Consider this case:

- Shuffle cycle set to 40
- At block height 38 a new Validator was added
- At block height 39 a new quorum is generated because the new Validator was added
- At block height 40 a shuffle occurs

It is unnecessary to shuffle the quorum at height 40, a new quorum was already generated at height 39. It should be shuffled +40 of the last quorum change height (which was 39, therefore 39 + 40 = 79, the next quorum shuffle).

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -66,6 +67,11 @@ public class Validator : FullNode, API
     /// Quorum generator parameters
     protected QuorumParams quorum_params;
 
+    /// The last height at which a quorum shuffle took place.
+    /// If a new block is externalized and `last_shuffle_height`
+    /// is >= `QuorumShuffleCycle` then the quorum will be reshuffled.
+    private Height last_shuffle_height = Height(0);
```
I don't think that comment is right, because what you are describing is a target, not a "last". Maybe missing an offset ?

AndrejMitrovic replied:
No I really am describing when the last shuffle occured.

AndrejMitrovic replied:
And it is implied that one happened in the genesis block, because that's the first time the quorum "changes".

Geod24 replied:
Hum... I see: `last_shuffle_height >= QuorumShuffleCycle`. I thought `QuorumShuffleCycle` was the config option, but it's the target ?

AndrejMitrovic replied:
I don't understand your terminologies here. Option vs Target.

We want there to be a quorum shuffle at least every X blocks, but they may happen more often if the set of enrollments has changed in the meantime. That's what I commented here https://github.com/bpfkorea/agora/pull/1072#issuecomment-669645266

Geod24 replied:
Mondays are hard.

`last_shuffle_height` is height at which the quorum was most recently shuffled;
`target_shuffle_height`, if it existed, would be the height at which the quorum is scheduled to be next shuffled;
`quorum_shuffle_cycle` is the difference between those two values;

AndrejMitrovic replied:
target would be a misnomer though.

If we have `quorum_shuffle_cycle` set to 10 it might seem a good idea to set `target_shuffle_height` to 10 at the genesis block 0. But if there is a new enrollment at block height 5 then at height 5 we will already shuffle the quorums, so then we would have to shift the target to `15`. 

It seemed unnecessary to me to introduce another state variable when we can just check when was the last time the quorum was shuffled and if the diff is `quorum_shuffle_cycle` then shuffle it again.

AndrejMitrovic replied:
I did just notice there is a bug in the logic. I was using `>` instead of `>=`.

I'll change the names of these variables.

AndrejMitrovic replied:
Ah the bug was that `quorum_shuffle_cycle` was not passed to `ConsensusParams`..

This default parameter stuff needs to be removed: https://github.com/bpfkorea/agora/blob/b602254d522f5bc2e5b5e96525c4d61e7c9180ff/source/agora/consensus/data/ConsensusParams.d#L51

It's just going to become a source of bugs.

Geod24 commented on source/agora/test/QuorumShuffle.d:
```diff
@@ -0,0 +1,469 @@
+            format("Node %s has block height %s. Expected: %s",
+                idx, node.getBlockHeight(), 3)));
+
+    enum quorums_2 = [
```
That seems like a common occurence... Maybe we should make this a bit easier / less invasive ?

AndrejMitrovic replied:
Hmm. I can try to calculate the preimage sum here and then call the quorum generator directly from the unittests and compare the quorums to see if they shuffled. It's definitely better than having to do these steps manually - thanks.

Geod24 commented on source/agora/test/QuorumShuffle.d:
```diff
@@ -0,0 +1,469 @@
+import std.range;
+
+import core.thread;
+import core.time;
```
Some of those imports should be useless

AndrejMitrovic replied:
Oh yes. Only 2 of them are actually used. Thanks.

Geod24 commented on source/agora/test/QuorumShuffle.d:
```diff
@@ -0,0 +1,469 @@
+    auto network = makeTestNetwork(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    //scope(failure) network.printLogs();
```
?

Geod24 commented on source/agora/test/QuorumShuffle.d:
```diff
@@ -0,0 +1,469 @@
+            WK.Keys.C.address]),
+    ];
+
+    version (none)
+    foreach (idx, node; nodes.enumerate)
+        writefln("Node %s: %s\n", idx, node.getQuorumConfig);
```
Leftover ?

AndrejMitrovic replied:
It's inside of `version (none)`, I left it in purposefully for when we change the quorum generator settings. Then you can uncomment `version (none)` and easily modify the test when the results are expected to change.

AndrejMitrovic replied:
Should I name it something like `version (PrintQuorums)`? I guess that could prevent needing to recompile.

AndrejMitrovic replied:
Prior art: https://github.com/Geod24/localrest/blob/37460b237c3bf14544dcd46d8d673bbe36d2d4db/source/geod24/LocalRest.d#L1437-L1438

Geod24 replied:
`version (none)` works for me, was just thinking you forgot them. I'm happy to keep them.

Geod24 commented on source/agora/test/QuorumShuffle.d:
```diff
@@ -0,0 +1,469 @@
+
+    void createBlock (ulong height)
+    {
+        txs = makeChainedTransactions(WK.Keys.Genesis, txs, 1);
```
`TxBuilder` ?

AndrejMitrovic replied:
Sure. I'll fix it.

Geod24 replied:
👀

AndrejMitrovic replied:
Oh I just realized this is a duplicate function of `makeBlock` above. Heh.

Geod24 commented on source/agora/test/QuorumShuffle.d:
```diff
@@ -0,0 +1,469 @@
+            WK.Keys.C.address]),
+    ];
+
+    version (none)
+    foreach (idx, node; nodes.enumerate)
+        writefln("Node %s: %s\n", idx, node.getQuorumConfig);
```
Leftover

AndrejMitrovic commented:
Thanks for the feedback, there were some nice ideas in here.

AndrejMitrovic commented:
I renamed the parameter to make it a little bit more obvious what it does. And I fixed the tests. The quorums are now all unique on every shuffle as expected.

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -192,7 +192,7 @@ public class Validator : FullNode, API
     {
         this.startPeriodicDiscovery();
-        this.taskman.setTimer(10.seconds, &this.checkRevealPreimage, Periodic.Yes);
+        this.taskman.setTimer(1.seconds, &this.checkRevealPreimage, Periodic.Yes);
```
Let's turn this into a node config option.

AndrejMitrovic commented:
Ah I see what the last problem remains with the CircleCI failure:

```
node-4_1_1bd1427fc5a0 | 2020-08-13 05:02:03,512 Info [agora.network.NetworkManager] - Found new Validator: http://node-7:7826 (key: GDNODE7J5EUK7T6HLEO2FDUBWZEXVXHJO7C4AF5VZAKZENGQ4WR3IX2U)
node-4_1_1bd1427fc5a0 | 2020-08-13 05:02:03,890 Info [agora.network.NetworkManager] - Found new Validator: http://node-3:3826 (key: GDNODE3EWQKF33TPK35DAQ3KXAYSOT4E4ACDOVJMDZQDVKP66IMJEACM)
node-4_1_1bd1427fc5a0 | 2020-08-13 05:02:04,335 Info [agora.network.NetworkManager] - Found new Validator: http://node-4:4826 (key: GDNODE4KTE7VQUHVBLXIGD7VEFY57X4XV547P72D37SDG7UEO7MWOSNY)
node-4_1_1bd1427fc5a0 | 2020-08-13 05:02:05,454 Info [agora.network.NetworkManager] - Found new Validator: http://node-2:2826 (key: GDNODE2IMTDH7SZHXWDS24EZCMYCEJMRZWB3S4HLRIUP6UNGKVVFLVHQ)
node-4_1_1bd1427fc5a0 | 2020-08-13 05:02:05,547 Info [agora.network.NetworkManager] - Found new Validator: http://node-6:6826 (key: GDNODE6ZXW2NNOOQIGN24MBEZRO5226LSMHGQA3MUAMYQSTJVR7XT6GH)
node-4_1_1bd1427fc5a0 | 2020-08-13 05:02:05,731 Info [agora.network.NetworkManager] - Found new FullNode: http://node-0:1826
Linking...
```

Node 4 is connecting to itself, but not to node 5. This is a symptom of https://github.com/bpfkorea/agora/issues/772.

Now the question is why didn't it connect to node 5. Did it stall somehow? I'll add some logging.

AndrejMitrovic commented:
```
node-5_1  | 2020-08-13 05:41:22,288 Info [agora.network.NetworkManager] - Todo addresses { _set: [ "http://172.20.0.4:39057": true, "http://172.20.0.7:48129": true, "http://172.20.0.7:57391": true, "http://172.20.0.8:58561": true, "http://172.20.0.5:58077": true, "http://172.20.0.8:54461": true, "http://172.20.0.7:43823": true, "http://172.20.0.2:47551": true, "http://172.20.0.6:42397": true, "http://172.20.0.3:37267": true, "http://172.20.0.2:55307": true ] }
```

~It looks like there is memory corruption coming from somewhere.~

Edit: it's not memory corruption.

AndrejMitrovic commented:
What we're actually seeing is that the node runs out of connection tasks (there is a preconfigured limit of 10). But the reason it runs out is because it's connecting to a very large set of IP & Port ranges. For example:

```
node-0_1  | 2020-08-13 05:56:03,518 Info [...] - registerListener: http://172.20.0.8:54471
node-0_1  | 2020-08-13 05:56:04,349 Info [...] - registerListener: http://172.20.0.7:34819
node-0_1  | 2020-08-13 05:56:04,682 Info [...] - registerListener: http://172.20.0.5:53267
node-0_1  | 2020-08-13 05:56:04,721 Info [...] - registerListener: http://172.20.0.6:36671
node-0_1  | 2020-08-13 05:56:04,730 Info [...] - registerListener: http://172.20.0.2:47281
node-0_1  | 2020-08-13 05:56:04,742 Info [...] - registerListener: http://172.20.0.3:44341
```

```
node-3_1  | 2020-08-13 05:56:04,722 Info [...] - registerListener: http://172.20.0.5:46665
node-3_1  | 2020-08-13 05:56:04,725 Info [...] - registerListener: http://172.20.0.6:36691
node-3_1  | 2020-08-13 05:56:04,730 Info [...] - registerListener: http://172.20.0.2:37641
node-3_1  | 2020-08-13 05:56:04,742 Info [...] - registerListener: http://172.20.0.3:34367
node-3_1  | 2020-08-13 05:56:04,981 Info [...] - registerListener: http://172.20.0.4:51453
node-3_1  | 2020-08-13 05:56:05,150 Info [...] - registerListener: http://172.20.0.8:38775
```

Notice the ports are different for the two nodes. So when the nodes share their set of known IPs to each other the total amount of IP & port combinations explodes. For example:

```diff
diff --git a/source/agora/network/NetworkManager.d b/source/agora/network/NetworkManager.d
index 61d3d1d27..6517e8067 100644
--- a/source/agora/network/NetworkManager.d
+++ b/source/agora/network/NetworkManager.d
@@ -401,6 +401,7 @@ public class NetworkManager
         this.discovery_task.add(node.client);
         this.metadata.peers.put(node.address);
         this.known_addresses.put(node.address);
+        log.info("Total known IPs: {}", this.known_addresses.length);
         this.connection_tasks.remove(node.address);

         this.registerAsListener(node.client);
@@ -474,6 +475,7 @@ public class NetworkManager

     public void registerListener (Address address)
     {
+        log.info("registerListener: {}", address);
         if (this.shouldEstablishConnection(address))
             this.addAddress(address);
     }
@@ -519,7 +521,10 @@ public class NetworkManager
         // reasoning: while *we* cannot establish a connection with
         // a node it's possible other nodes in the network might be able to.
         if (!this.isOurOwnAddress(address))
+        {
             this.known_addresses.put(address);
+            log.info("Total known IPs: {}", this.known_addresses.length);
+        }
     }

     /***************************************************************************
```

And the result:

```
node-5_1  | 2020-08-13 06:03:56,789 Info [agora.network.NetworkManager] - Total known IPs: 51
```

I think we need to fix this code: https://github.com/bpfkorea/agora/blob/a490dade705e5d9367e74c40793f2f430ae52b76/source/agora/node/Runner.d#L79

But I'm not sure how to proceed.

AndrejMitrovic commented:
One way to fix this is to not add the client's IP & Port to the known addresses.

Geod24 commented:
(Rebased)

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -836,6 +837,9 @@ public interface TestAPI : ValidatorAPI
     public void registerListenerAddress (Address address);
+
+    /// Get the list of expected quorum configs
+    public QuorumConfig[] getExpectedQuorums (in PublicKey[], Height);
```
Should document parameters

Geod24 commented on source/agora/test/QuorumShuffle.d:
```diff
@@ -0,0 +1,203 @@
+
+    void makeBlock (ulong height)
+    {
+        txs = txs.filter!(tx => tx.type == TxType.Payment)
+            .map!(tx => iota(tx.outputs.length)
+                .map!(idx => TxBuilder(tx, cast(uint)idx)))
+            .joiner().takeExactly(8)
+            .map!(txb => txb.refund(WK.Keys.Genesis.address).sign()).array;
```
- Use `spendable()`;
- Don't use `Genesis`

AndrejMitrovic replied:
What do you mean by "don't use Genesis". The special Genesis address?

Geod24 replied:
> The special Genesis address?

Yes

AndrejMitrovic replied:
`spendable` requires us to do a lot of back and forth btw - I can see this is going to become another code segment that is copy-pasted everywhere:


```D
const(Block)[] blocks = [network.blocks[$ - 1]];

void makeBlock (ulong height)
{
    auto txs = blocks[$ - 1].spendable.takeExactly(8)
        .map!(txb => txb.refund(WK.Keys.Genesis.address).sign()).array;
    txs.each!(tx => nodes[0].putTransaction(tx));
    nodes.enumerate.each!((idx, node) =>
        retryFor(node.getBlockHeight() == height, 5.seconds,
            format("Node %s has block height %s. Expected: %s",
                idx, node.getBlockHeight(), height)));
    retryFor((blocks = nodes[0].getBlocksFrom(height, 1)).length == 1, 5.seconds,
            format("Node 0 getBlocksFrom(%s, 1) failed", height));
}
```

Geod24 replied:
1) Do we need `takeExactly` since we're transitioning away from tx-count-based block creation ?
2) Why do you set the refund address ? `sign()` is enough.
3) Since the transactions are  not  used as an array anywhere, no need to `array` then `each`, just `each` directly.

In short:
```D
blocks[$ - 1].spendable.each!(txb => nodes[0].putTransaction(txb.sign()));
```

AndrejMitrovic replied:
Oh sorry I didn't realize `sign` does this already.

> Do we need takeExactly since we're transitioning away from tx-count-based block creation ?

Well we're still transitioning. :) 

Once we transition we can get rid of that, sure.

Geod24 replied:
> Oh sorry I didn't realize `sign` does this already.

Yes, the only requirement for `TxBuilder` is a refund address. If you provide it with a single output, it'll use the beneficiary as refund address, and if you provide it with a transaction, it'll use the first output's beneficiary by default.

Geod24 reviewed (approved):
Just a few comments, otherwise LGTM. Feel free to merge after they are addressed.
